### ISSUE TITLE
CQL to ELM translator, how to obtain "codeProperty" for certain queries

### ISSUE THREAD

**Issue by amarvakul** [cqframework/clinical_quality_language#546]
CQL to ELM translator, "codeProperty" is not returned for certain queries

For following CQL Examples, codeProperty is not returned

```
define ClaimQuery1:  [Condition] C where C.code in Diabetes

define ClaimRetrieve2:  [Claim] C where exists C.item CI where (CI.revenue in  "Hospice Encounter")

```

Corresponding ELMs

> <def name="ClaimQuery1" context="Patient" accessLevel="Public">
         <resultTypeSpecifier xsi:type="ListTypeSpecifier">
            <elementType name="fhir:Condition" xsi:type="NamedTypeSpecifier"/>
         </resultTypeSpecifier>
         <expression xsi:type="Query">
            <resultTypeSpecifier xsi:type="ListTypeSpecifier">
               <elementType name="fhir:Condition" xsi:type="NamedTypeSpecifier"/>
            </resultTypeSpecifier>
            <source alias="C">
               <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                  <elementType name="fhir:Condition" xsi:type="NamedTypeSpecifier"/>
               </resultTypeSpecifier>
               <expression dataType="fhir:Condition" xsi:type="Retrieve">
                  <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                     <elementType name="fhir:Condition" xsi:type="NamedTypeSpecifier"/>
                  </resultTypeSpecifier>
               </expression>
            </source>
            <where resultTypeName="t:Boolean" xsi:type="InValueSet">
               <signature name="t:Concept" xsi:type="NamedTypeSpecifier"/>
               <code name="ToConcept" libraryName="FHIRHelpers" xsi:type="FunctionRef">
                  <signature name="fhir:CodeableConcept" xsi:type="NamedTypeSpecifier"/>
                  <operand resultTypeName="fhir:CodeableConcept" path="code" scope="C" xsi:type="Property"/>
               </code>
               <valueset name="Diabetes">
                  <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                     <elementType name="t:Code" xsi:type="NamedTypeSpecifier"/>
                  </resultTypeSpecifier>
               </valueset>
            </where>
         </expression>
      </def>

> <def name=ClaimRetrieve2 context="Patient" accessLevel="Public">
         <resultTypeSpecifier xsi:type="ListTypeSpecifier">
            <elementType name="fhir:Claim" xsi:type="NamedTypeSpecifier"/>
         </resultTypeSpecifier>
         <expression xsi:type="Query">
            <resultTypeSpecifier xsi:type="ListTypeSpecifier">
               <elementType name="fhir:Claim" xsi:type="NamedTypeSpecifier"/>
            </resultTypeSpecifier>
            <source alias="C">
               <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                  <elementType name="fhir:Claim" xsi:type="NamedTypeSpecifier"/>
               </resultTypeSpecifier>
               <expression dataType="fhir:Claim" xsi:type="Retrieve">
                  <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                     <elementType name="fhir:Claim" xsi:type="NamedTypeSpecifier"/>
                  </resultTypeSpecifier>
               </expression>
            </source>
            <where resultTypeName="t:Boolean" xsi:type="Exists">
               <signature xsi:type="ListTypeSpecifier">
                  <elementType name="fhir:Claim.Item" xsi:type="NamedTypeSpecifier"/>
               </signature>
               <operand xsi:type="Query">
                  <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                     <elementType name="fhir:Claim.Item" xsi:type="NamedTypeSpecifier"/>
                  </resultTypeSpecifier>
                  <source alias="CI">
                     <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                        <elementType name="fhir:Claim.Item" xsi:type="NamedTypeSpecifier"/>
                     </resultTypeSpecifier>
                     <expression path="item" scope="C" xsi:type="Property">
                        <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                           <elementType name="fhir:Claim.Item" xsi:type="NamedTypeSpecifier"/>
                        </resultTypeSpecifier>
                     </expression>
                  </source>
                  <where resultTypeName="t:Boolean" xsi:type="InValueSet">
                     <signature name="t:Concept" xsi:type="NamedTypeSpecifier"/>
                     <code name="ToConcept" libraryName="FHIRHelpers" xsi:type="FunctionRef">
                        <signature name="fhir:CodeableConcept" xsi:type="NamedTypeSpecifier"/>
                        <operand resultTypeName="fhir:CodeableConcept" path="revenue" scope="CI" xsi:type="Property"/>
                     </code>
                     <valueset name="Hospice Encounter">
                        <resultTypeSpecifier xsi:type="ListTypeSpecifier">
                           <elementType name="t:Code" xsi:type="NamedTypeSpecifier"/>
                        </resultTypeSpecifier>
                     </valueset>
                  </where>
               </operand>
            </where>
         </expression>
      </def>

We tried to write like below, but this throws ELM error, how do we write so that we can get codeProperty and also the CQL is valid.

`define ClaimRetrieve3: [Claim: item.revenue in "Hospice Encounter"]
`

Expectations: "codeProperty" to be returned. For ClaimRetrieve2 example, "codeProperty" to be returned at the top level query or in a completely separate structure outside of the  "statements" in the ELM. As we are parsing these optimization hints and push them as predicates at the time of data access of a resource otherwise we would not have these hints information at the right if we only find them at the sub-query.

**Comment by brynrhodes** [comment]
Note that this is currently being worked on as a $datarequirements evaluation component in the CQF Tooling repository.

**Comment by brynrhodes** [comment]
This is addressed by the new DataRequirementsProcessor included in the 1.5.4 release, in that the data requirements returned from that analysis will include the code property for both of the requested queries (and several other patterns as well).
